external/fv3fit/fv3fit/_shared/xr_prediction.py

```diff
@@ -70,3 +77,86 @@ def predict_on_dataset(
         unstacked_dims=unstacked_dims,
     )
     return match_prediction_to_input_coords(X, return_ds)
+
+
+@io.register("dataset-predictor")
+class DatasetPredictor(Predictor):
+    """
+    Base model is an ArrayPredictor which takes in a sequence of
+    input arrays and outputs a sequence of output arrays.
+    Normalization is contained within the base ArrayPredictor.
+
+    Assumes wrapped model accepts [sample, feature] arrays.
+    """
+
+    _BASE_MODEL_NAME = "base_model"
+    _CONFIG_FILENAME = "config.yaml"
+
+    def __init__(
+        self,
+        input_variables: Iterable[Hashable],
+        output_variables: Iterable[Hashable],
+        model: ArrayPredictor,
+        unstacked_dims: Sequence[str],
+        n_halo: int = 0,
+    ):
+        """Initialize the predictor
+
+        Args:
+            input_variables: names of input variables
+            output_variables: names of output variables
+            model: base model to wrap. Must be in fv3fit.io registry
+            unstacked_dims: non-sample dimensions of model output
+            n_halo: number of halo points required in input data
+        """
+        super().__init__(input_variables, output_variables)
+        self.input_variables = input_variables
+        self.output_variables = output_variables
+        self.model = model
+        self._n_halo = n_halo
+        self._unstacked_dims = unstacked_dims
+
+    def predict(self, X: xr.Dataset) -> xr.Dataset:
+        """Predict an output xarray dataset from an input xarray dataset."""
+        return predict_on_dataset(
+            model=self.model,
+            X=X,
+            input_variables=self.input_variables,
+            output_variables=self.output_variables,
+            n_halo=self._n_halo,
+            unstacked_dims=self._unstacked_dims,
+        )
+
+    @classmethod
+    def load(cls, path: str) -> "DatasetPredictor":
+        """Load a serialized model from a directory."""
+        with get_dir(path) as path:
+            base_model_filename = os.path.join(path, cls._BASE_MODEL_NAME)
+            base_model: ArrayPredictor = io.load(base_model_filename)  # type: ignore
```

Comment:
`ArrayPredictor` is a `Reloadable`, I'm not sure why mypy still complains here.

Reply:
`io.load` returns a `Reloadable`, which isn't guarunteed to be an `ArrayPredictor` from types alone. Only certain string values will give an `ArrayPredictor`. For type safety you should directly call `ArrayPredictor.load` instead of `io.load` if this truly is supposed to be an ArrayPredictor always, or type this as `Reloadable` if not.

I'm a little confused because I don't see where you defined `dump` and `load` on `ArrayPredictor`, but if you don't define them it should cause an error when initialized. Did you add a test that uses this class?

Reply:
Or yeah, make `ArrayPredictor` another subclass of `Reloadable` that is like `Predictor` but does the type of predictions you want here.

Reply:
Predictor is a dataset predictor, it's predict api takes in an xr.Dataset. If you like you could rename Predictor to `DatasetPredictor` to make it clear how it relates to `ArrayPredictor`. Given 

Reloadable is just a protocol saying that the thing you return from the train function must be dumpable and loadable, it doesn't implement those methods.

I would strongly encourage implementing as little functionality as you can in an abstract base class, so that the ABC acts only as an interface/defines what should be implemented, and the code implementation itself doesn't use inheritance/isn't split between two classes.

It sounds like what you want to make is an analog to PureKerasModel, but one that wraps an ArrayPredictor to make a Predictor. You probably want to make a concrete implementation of Predictor for this, instead of defining another ABC/interface. If you do that, its functionality/code can be tested the same way we test PureKerasModel.

Reply:
After discussion with Jeremy and Brian, I changed the `ArrayPredictor` to be an abstract base class and will keep the current usage of `io.load/dump` in the `DatasetPredictor`'s dump/load methods (since this class will be used for multiple types of ArrayPredictors). 

I did look at unifying these changes with `PureKerasModel` but realized that class has its own additional methods like `input_sensitivty, get_dict_compatible_model` that are very specific to having a keras model attribute, so I am leaving this new `DatasetPredictor` as a separate class. Both `PureKerasModel` and `DatasetPredictor` do at least make use of the same `predict_on_dataset` function that was refactored out in a previous PR.